src/handlers.rs

```diff
@@ -68,3 +72,111 @@ pub async fn decode(req: web::Json<Request>) -> Result<impl Responder> {
     };
     Ok(web::Json(response))
 }
+
+async fn get_tx_logs(txn_hash: &DisplayBytes, network: &String) -> Result<Vec<TxLog>> {
+    let res = reqwest::get(format!(
+        "https://blockscout.com/{}/api?module=transaction&action=gettxinfo&txhash={}",
+        network, txn_hash
+    ))
+    .await
+    .map_err(|err| anyhow!(err))?;
+
+    if res.status() == StatusCode::NOT_FOUND {
+        return Err(Error::GetTxInfo("Wrong network".to_string()));
+    }
+
+    let res: Transaction = res.json().await.map_err(|err| anyhow!(err))?;
+
+    if res.status != "1" {
+        return Err(Error::GetTxInfo(res.message));
+    }
+
+    res.result
+        .map(|result| result.logs)
+        .ok_or_else(|| Error::Other(anyhow!("Missing input block")))
+}
+
+async fn get_contract_abi(network: &String, address: &DisplayBytes) -> Result<Contract> {
+    let res = reqwest::get(format!(
+        "https://blockscout.com/{}/api?module=contract&action=getabi&address={}",
+        network, address
+    ))
+    .await
+    .map_err(|err| anyhow!(err))?;
+
+    if res.status() == StatusCode::NOT_FOUND {
+        return Err(Error::GetTxInfo("Wrong network".to_string()));
+    }
+
+    let res: AbiResponse = res.json().await.map_err(|err| anyhow!(err))?;
+
+    if res.status == "0" {
+        return Err(Error::GetTxInfo(res.message));
+    }
+
+    let c: Contract = serde_json::from_str(
+        &res.result
+            .ok_or_else(|| anyhow!("Abi parsing failed".to_string()))?,
+    )
+    .map_err(|err| anyhow!(err))?;
+
+    Ok(c)
+}
+
+async fn get_log_event(
+    network: &String,
+    log: &TxLog,
+    abi_map: &mut HashMap<Vec<u8>, Contract>,
+) -> Result<Event> {
+    let abi = abi_map
+        .entry(log.address.to_vec())
+        .or_insert(get_contract_abi(network, &log.address).await?);
+
+    for event in abi.events() {
+        if log.topics[0]
+            .as_ref()
+            .ok_or_else(|| anyhow!("Failed to decode anonymous event".to_string()))?
+            .0
+            == event.signature().as_bytes()
+        {
+            return Ok(event.clone());
+        }
+    }
+
+    Err(Error::Other(anyhow!("Failed to find event".to_string())))
+}
+
+async fn process_events(network: &String, logs: &Vec<TxLog>) -> Result<Vec<(Event, Log, String)>> {
+    let mut abi_map: HashMap<Vec<u8>, Contract> = HashMap::new();
+
+    let mut processed_events: Vec<(Event, Log, String)> = Vec::new();
```

Comment:
I don't insist on this variant, but you could do this with iterators like something
```rust
let processed_events: Result<_> = logs
 .into_iter()
 .filter_map(|log| 
    get_log_event(network, log, &mut abi_map),map(|event| (event, log)).ok())
 .map(|(event, log)|  event.parse_log(...).map(|parsed_log| (event, parsed_log, log)).map_err(..))
 .collect()
let processed_events = processed_events?;
```